### DISCUSSION TITLE
Warning: should not be used directly in a SQL statement context, such as passing to methods such as session.execute().  This usage will be disallowed in a future release.  Please use Core select() / update() / delete() etc. with Session.execute() and other statement execution methods.

### DISCUSSION THREAD

**Discussion by swills1** [sqlalchemy/sqlalchemy#8312] [Usage Questions] [answered]
Hello.

I have this code.
```
with Session(config.connect()) as session:
    test = session.query(Tag.tag, func.count('Tag.tag').label('Label')).group_by(Tag.tag)
    result = session.execute(test).fetchall()
    print(result)
```

and it works, but gives this  warning,

```
SADeprecationWarning: Object <sqlalchemy.orm.query.Query object at 0x7f5269545600> should not be used directly in a SQL statement context, such as passing to methods such as session.execute().  This usage will be disallowed in a future release.  Please use Core select() / update() / delete() etc. with Session.execute() and other statement execution methods.
  result = session.execute(test).fetchall()
```

What is the proper way I am supposed to be doing the above?

**Comment by CaselIT** [comment]
Query should not be passed to execute. Above you can do `test.all()`

**Reply by swills1** [reply]
Ah, so I should be doing this then, correct?

```
with Session(config.connect()) as session:
    test = session.query(Tag.tag, func.count('Tag.tag').label('label')).group_by(Tag.tag)
    result = (test).all()
   ```
I tested and it seems to work. 

Thanks. Is there documentation where I can read about this change overall? Thanks again for the insight.

**Reply by CaselIT** [reply]
There was no charge. The query has always worked this way, it was an oversight allowing execute to work with a query. What has changed was the addition of the warning.
You can look at the query documentation on the docs https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query#sqlalchemy.orm.Query

Also note that query is legacy, so you may what to consider using select instead. Take a look at https://docs.sqlalchemy.org/en/14/tutorial/index.html for details

**Reply by swills1** [reply]
Yes, I understand query is legacy.

I think we may be on different pages. I just want to know the proper way of doing the above with the current version of SQLAlchemy. Not simply modify that exact  code to work in general. 

The warning indicates that I _should_ still be using `execute`, but with select instead. Looking at the documentation you linked, it appears the proper way to do the above with the current version of SA would be;

```
test = select(JokeTag.tag, func.count('JokeTag.tag').label('jokes')).group_by(JokeTag.tag)
with Session(config.connect()) as session:
    test = session.execute(test)
    for row in test:
        print(row)
```

Is that correct?

Source: (https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#the-select-sql-expression-construct)

Thanks again.

**Reply by CaselIT** [reply]
I would not call query "not correct". So using `query(...).all()` is correct. Otherwise as you mentioned using select is also correct and is the suggested way of using sqlalchemy for new code.

To get all values you can call `.all()` on the `execute` result.

**Reply by swills1** [reply]
"Is that correct" referred to - "Is my statement correct?" Not, is one methodology correct and the other incorrect. You'll notice I never used the word "correct" to refer to a methodology. I intentionally used the word **proper**, any time I refereed to a methodology.

According to you, _query()_ is being deprecated. If it is being deprecated, it would be ill advised to use it for any new projects. 

So based on what you have said and what the docs say - logic would dictate, this is the _proper_ way;

https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#the-select-sql-expression-construct

Thank you.

**Reply by CaselIT** [reply]
> According to you, _query()_ is being deprecated. If it is being deprecated, it would be ill advised to use it for any new projects.

Note that legacy does not mean depracated. There are currently no plan to deprecate it, but it will likely not receive any new feature.

> So based on what you have said and what the docs say - logic would dictate, this is the _proper_ way;

I would not call query improper. Using select is the suggested way and what should be used in new projects, but query is not improper.

In any case to summarize:
- if you are writing new code select is probably preferable
- there is no need to change all existing usage of query in existing code
- in all cases query should not be passed to `execute()`

**Reply by swills1** [reply]
Thanks!

**Reply by zzzeek** [reply]
whew!

**Reply by swills1** [reply]
I'm going to donate as appreciation for the help.

**Comment by swills1** [comment, marked as answer]
This is the answer. Thanks @CaselIT 

https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#the-select-sql-expression-construct
```
stmt = select(User).where(User.name == 'spongebob')
with Session(engine) as session:
    for row in session.execute(stmt):
        print(row)
```
